File: 1851-minimum-interval-to-include-each-query/1851-minimum-interval-to-include-each-query.py
import heapq

class Solution:
    def minInterval(self, intervals, queries):
        result = {}
        sortedQueries = sorted(queries)
        sortedIntervals = sorted(intervals)
        pq = []
        i = 0
        n = len(intervals)
        
        for q in sortedQueries:
            while i < n and sortedIntervals[i][0] <= q:
                start, end = sortedIntervals[i][0], sortedIntervals[i][1]
                heapq.heappush(pq, (end - start + 1, end))
                i += 1
            
            while pq and pq[0][1] < q:
                heapq.heappop(pq)
            
            result[q] = pq[0][0] if pq else -1
        
        return [result[q] for q in queries]

File: 1851-minimum-interval-to-include-each-query/1851-minimum-interval-to-include-each-query.java
import java.util.*;

class Solution {
    static class Comp implements Comparator<Pair<Integer, Integer>> {
        public int compare(Pair<Integer, Integer> p1, Pair<Integer, Integer> p2) {
            return p1.getKey().compareTo(p2.getKey());
        }
    }

    public int[] minInterval(int[][] intervals, int[] queries) {
        PriorityQueue<Pair<Integer, Integer>> pq = new PriorityQueue<>(new Comp());
        Map<Integer, Integer> result = new HashMap<>();
        List<Integer> sortedQueries = new ArrayList<>();
        for (int q : queries) {
            sortedQueries.add(q);
        }
        Collections.sort(sortedQueries);
        List<Pair<Integer, Integer>> sortedIntervals = new ArrayList<>();
        for (int[] interval : intervals) {
            sortedIntervals.add(new Pair<>(interval[0], interval[1]));
        }
        Collections.sort(sortedIntervals, new Comp());

        int i = 0;
        int n = intervals.length;
        for (int q : sortedQueries) {
            while (i < n && sortedIntervals.get(i).getKey() <= q) {
                int start = sortedIntervals.get(i).getKey();
                int end = sortedIntervals.get(i).getValue();
                pq.offer(new Pair<>(end - start + 1, end));
                i++;
            }
            while (!pq.isEmpty() && pq.peek().getValue() < q) {
                pq.poll();
            }
            result.put(q, pq.isEmpty() ? -1 : pq.peek().getKey());
        }

        int[] res = new int[queries.length];
        for (int j = 0; j < queries.length; j++) {
            res[j] = result.get(queries[j]);
        }
        return res;
    }
}
